libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/InMemoryDataStore.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Override
+    public T getItem(String partitionValue) {
+        return records.get(getKey(partitionValue, null));
+    }
+
+    @Override
+    public T getItem(String partitionValue, boolean warnOnNull) {
+        return getItem(partitionValue);
+    }
+
+    @Override
+    public T getItemByIndex(String indexName, String value) {
+        Method indexMethod = null;
+        for (var method : klass.getMethods()) {
+            if (method.isAnnotationPresent(DynamoDbSecondaryPartitionKey.class)) {
+                var annotation = method.getAnnotation(DynamoDbSecondaryPartitionKey.class);
+                if (Arrays.asList(annotation.indexNames()).contains(indexName)) {
+                    indexMethod = method;
+                }
+            }
+        }
+        if (indexMethod == null) {
+            throw new IllegalArgumentException("Missing index " + indexName);
+        }
+
+        for (var item : records.values()) {
+            try {
+                if (indexMethod.invoke(item).toString().equals(value)) {
+                    return item;
+                }
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new IllegalArgumentException("Could not access index " + indexName);
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public List<T> getItems(String partitionValue) {
+        return records.values().stream()
+                .filter(i -> getPartitionKey(i).equals(partitionValue))
+                .toList();
+    }
+
+    @SuppressWarnings(
+            "java:S3011") // We rely on field access to avoid needing to search for a getter
+    @Override
+    public List<T> getItemsWithBooleanAttribute(String partitionValue, String name, boolean value) {
+        try {
+            var field = klass.getDeclaredField(name);
+            field.setAccessible(true);
+            return records.values().stream()
+                    .filter(
+                            i -> {
+                                try {
+                                    return Boolean.valueOf(value).equals(field.get(i));
+                                } catch (IllegalAccessException e) {
+                                    throw new IllegalArgumentException(
+                                            "Could not access attribute " + name, e);
+                                }
+                            })
+                    .toList();
+        } catch (NoSuchFieldException e) {
+            throw new IllegalArgumentException("Could not find boolean attribute " + name, e);
+        }
+    }
+
+    @Override
+    public List<T> getItemsBySortKeyPrefix(String partitionValue, String sortPrefix) {
+        return records.values().stream()
+                .filter(
+                        i ->
+                                getPartitionKey(i).equals(partitionValue)
+                                        && getSortKey(i).startsWith(sortPrefix))
```

Comment:
`getSortKey()` can return null, which will blow this up? Hopefully we wouldn't use this method on a table that doesn't have a sort key, but might be worth addressing.